c++/src/capnp/rpc-twoparty.c++

```diff
@@ -248,52 +246,31 @@ kj::Own<OutgoingRpcMessage> TwoPartyVatNetwork::newOutgoingMessage(uint firstSeg
 }
 
 kj::Promise<kj::Maybe<kj::Own<IncomingRpcMessage>>> TwoPartyVatNetwork::receiveIncomingMessage() {
-  return kj::evalLater([this]() {
-    KJ_SWITCH_ONEOF(stream) {
-      KJ_CASE_ONEOF(ioStream, kj::AsyncIoStream*) {
-        return tryReadMessage(*ioStream, receiveOptions)
-            .then([](kj::Maybe<kj::Own<MessageReader>>&& message)
-                  -> kj::Maybe<kj::Own<IncomingRpcMessage>> {
-          KJ_IF_MAYBE(m, message) {
-            return kj::Own<IncomingRpcMessage>(kj::heap<IncomingMessageImpl>(kj::mv(*m)));
-          } else {
-            return nullptr;
-          }
-        });
-      }
-      KJ_CASE_ONEOF(capStream, kj::AsyncCapabilityStream*) {
-        auto fdSpace = kj::heapArray<kj::AutoCloseFd>(maxFdsPerMessage);
-        auto promise = tryReadMessage(*capStream, fdSpace, receiveOptions);
-        return promise.then([fdSpace = kj::mv(fdSpace)]
-                            (kj::Maybe<MessageReaderAndFds>&& messageAndFds) mutable
-                          -> kj::Maybe<kj::Own<IncomingRpcMessage>> {
-          KJ_IF_MAYBE(m, messageAndFds) {
-            if (m->fds.size() > 0) {
-              return kj::Own<IncomingRpcMessage>(
-                  kj::heap<IncomingMessageImpl>(kj::mv(*m), kj::mv(fdSpace)));
-            } else {
-              return kj::Own<IncomingRpcMessage>(kj::heap<IncomingMessageImpl>(kj::mv(m->reader)));
-            }
-          } else {
-            return nullptr;
-          }
-        });
+  return kj::evalLater([this]() -> kj::Promise<kj::Maybe<kj::Own<IncomingRpcMessage>>> {
+    auto fdSpace = (maxFdsPerMessage == 0)
```

Comment:
Type deduction through ternary operators is a weird corner of C++ that I've grown to distrust, so I would probably have written this as:

```
kj::Array<kj::AutoCloseFd> fdSpace = nullptr;
if (maxFdsPerMessage > 0) {
  fdSpace = kj::heapArray<kj::AutoCloseFd>(maxFdsPerMessage);
}
```

I realize that probably looks pretty awful to a Haskell programmer... :)

The alternative that I trust is to make sure both branches resolve to the same type, but that'll end up even more verbose in this case.